torchdata/dataloader2/dataloader2.py

```diff
@@ -169,14 +169,18 @@ def shutdown(self) -> None:
         r"""
         Shuts down ``ReadingService`` and clean up iterator.
         """
-        if not self._reset_iter:
-            self._reset_iter = True
-            self._datapipe_iter = None
-        if not self._terminated:
-            if self.reading_service is not None:
-                self.reading_service.finalize_iteration()
-                self.reading_service.finalize()
-            self._terminated = True
+        try:
+            if not self._reset_iter:
+                self._reset_iter = True
+                self._datapipe_iter = None
+            if not self._terminated:
+                if self.reading_service is not None:
+                    self.reading_service.finalize_iteration()
+                    self.reading_service.finalize()
+                self._terminated = True
+        # Ignore AttributeError in case any attribute has been removed before `__del__`
+        except AttributeError:
+            pass
```

Comment:
Just to make sure I'm understanding correctly, seems like this was always needed?

Reply:
Yea, as long as we are accessing attributes within `__del__`, we should handle such scenario.